--- entrypoint-org	2017-05-18 04:42:48.000000000 +0100
+++ entrypoint-patched	2017-05-18 04:45:11.000000000 +0100
@@ -18,5 +18,10 @@
   cmp -s "${CA_CERTIFICATES_PATH}" "${LOCAL_CA_PATH}" || update_ca
 fi
 
+if [ ! -f "${CONFIG_FILE}" ]; then
+  # create the configuration file from the environment
+  gitlab-runner register -n
+fi
+
 # launch gitlab-ci-multi-runner passing all arguments
 exec gitlab-ci-multi-runner "$@"
